orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py

```diff
@@ -537,3 +533,51 @@ def __setstate__(self, state):
         self.__dict__.update(state)
 
         _utils.reinitialize_graph_execution_manager(self)
+
+    def _enable_conditional_optimizations(self, graph_transformer_config, inputs, kwargs):
+        """
+        Based on runtime inspection, enable conditional optimizations if applicable.
+
+        Input sparsity-based optimization workflows:
+        1. Input density observer is initialized when input density observer is available.
+        2. If compute optimizer is enabled, input density observer inspects input tensors and returns sparsity results.
+        3. If label or embedding input sparsity is found, graph transformer config is updated to enable sparsity-based
+           optimization.
+
+        """
+
+        # Set up data sparsity inspection if input density observer is available.
+        if self._rt_inspector.input_density_ob is not None:
+            self._rt_inspector.input_density_ob.initialize(
+                self._onnx_models.exported_model, self._graph_builder.get_graph_info().user_input_names
+            )
+
+            # Enable sparsity-based optimization when applicable.
+            # Be noted: this optimization is enabled when input density observer is available and compute_optimizer is
+            # enabled.
+            if self._enable_compute_optimizer:
+                detected_device = _utils.get_device_from_module(self._original_module) or _utils.get_device_from_inputs(
+                    inputs, kwargs
+                )
+
+                with _logger.suppress_os_stream_output(log_level=self._debug_options.logging.log_level):
+                    _, embed_sparsity_results, label_sparsity_results = _io._combine_input_buffers_initializers(
+                        self._graph_initializers,
+                        self._graph_builder.get_graph_info().user_input_names,
+                        self._input_info,
+                        self._flattened_module.named_buffers(),
+                        inputs,
+                        kwargs,
+                        detected_device,
+                        self._rt_inspector.input_density_ob,
+                    )
+
+                if len(label_sparsity_results) > 0:
+                    graph_transformer_config.sparse_label_input_names = label_sparsity_results
+                    warnings.warn(f"Label sparsity based optimization is on for {label_sparsity_results}", UserWarning)
+
+                if len(embed_sparsity_results) > 0:
+                    graph_transformer_config.sparse_embedding_input_names = embed_sparsity_results
+                    warnings.warn(
+                        f"Embedding sparsity based optimization is on for {embed_sparsity_results}", UserWarning
```

Comment:
Same here. Do we need this warning to the user? Considering that it is enabled by default?